rfcs/search_relevancy_sandbox/20230331-project_proposal_search_relevancy_sandbox.md

```diff
@@ -0,0 +1,145 @@
+
+- [ ] dhruvkb
+- [ ] sarayourfriend
+
+## Project summary
+
+<!-- A brief one or two sentence summary of the project's features -->
+
+Develop mechanisms and workflows for sustainably creating & updating our staging
+API database & Elasticsearch cluster, in order to enable testing changes to
+Elasticsearch indices. The staging setup will reflect production in two ways:
+proportionally by provider and with production-level volumes.
+
+## Goals
+
+<!-- Which yearly goal does this project advance? -->
+
+Yearly goal: **Result Relevancy**
+
+## Requirements
+
+<!-- Detailed descriptions of the features required for the project. Include user stories if you feel they'd be helpful, but focus on describing a specification for how the feature would work with an eye towards edge cases. -->
+
+The following should be available once this project is complete:
+
+1. A mechanism by which maintainers can easily update the staging API database
+   with recent data.
+2. A mechanism by which maintainers can create/update a small but
+   proportional-to-production-by-provider index.
+3. A mechanism by which maintainers can create/update a production-data-volume
+   sized index.
+4. A mechanism by which maintainers can easily deploy & iterate on new
+   Elasticsearch indexing configurations
+5. A mechanism by which maintainers can easily point the staging Elasticsearch
+   index aliases (currently `image` and `audio`) to one of the above.
+
+Many of these mechanisms will likely be built in Airflow.
+
+### Proportional by provider
+
+The proportional-by-provider index will be a subset of the production data,
+constructed so that the number of records by provider is roughly proportional to
+the percentages per provider that exist in production. This index will allow us
+to create tests down the line which might assess result relevancy in an
+automated way, i.e. based on how many results from a particular provider appear
+in results (see the
+[provider tally script](https://github.com/WordPress/openverse/pull/397) for
+examples of these results). This could be done by:
+
+1. Querying the `/stats` endpoint of the production API (e.g.
+   https://api.openverse.engineering/v1/images/stats/).
+2. Calculating percentage-per-provider.
+3. Computing the number of results per provider based on these percentages based
+   on the condition that the smallest provider has a minimum `N` results (e.g. a
+   minimum of 10 results).
+4. Pulling a number of results based on the previous step per-provider either
+   into a Postgres table or an Elasticsearch index (the specifics of this step
+   will be left for the implementation plan).
+
+This index will allow us to assess how changes to Elasticsearch index or the API
+will affect results using a smaller total result count. It will also allow us to
+test API results from all currently available providers.
+
+### Production data volume
+
+The production-data-volume index will be an index of similar size to production.
```

Comment:
What does "similar size" mean, in this context? Is the variation down to the fact that any data refresh started _after_ another will have slightly more data due to ongoing ingestion into the catalogue?

Reply:
Yes, that's exactly what I mean but I can be more specific. While I'm glad that the 500M document index worked well for your testing, I think it should be possible for us to match production close enough that staging trails it by a week (and at most a month's) worth of data. I will codify my intentions here a bit better.

Reply:
Makes sense, I don't think it is a huge distinction, I just wanted to make sure I was understanding what the possible discrepancy would be. Using production snapshots as described in the first implementation plan makes a lot of sense and clarifies the data lag that would appear when a data refresh runs.